src/libraries/System.Memory/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -588,52 +611,76 @@ private static unsafe void Ssse3Decode(ref byte* srcBytes, ref byte* destBytes,
             {
                 AssertRead<Vector128<sbyte>>(src, srcStart, sourceLength);
-                Vector128<sbyte> str = Sse2.LoadVector128(src).AsSByte();
+                Vector128<byte> str = Vector128.LoadUnsafe(ref *src);
 
                 // lookup
-                Vector128<sbyte> hiNibbles = Sse2.And(Sse2.ShiftRightLogical(str.AsInt32(), 4).AsSByte(), mask2F);
-                Vector128<sbyte> loNibbles = Sse2.And(str, mask2F);
-                Vector128<sbyte> hi = Ssse3.Shuffle(lutHi, hiNibbles);
-                Vector128<sbyte> lo = Ssse3.Shuffle(lutLo, loNibbles);
+                Vector128<byte> hiNibbles = Vector128.ShiftRightLogical(str.AsInt32(), 4).AsByte();
+                if (Ssse3.IsSupported)
+                {
+                    // Ensure the high bits are masked off for the shuffle.
+                    hiNibbles = Vector128.BitwiseAnd(hiNibbles, mask2F);
+                }
+                Vector128<byte> hi = SimdShuffle(lutHi, hiNibbles, f);
+                Vector128<byte> lo = SimdShuffle(lutLo, str, f);
 
                 // Check for invalid input: if any "and" values from lo and hi are not zero,
                 // fall back on bytewise code to do error checking and reporting:
-                if (Sse2.MoveMask(Sse2.CompareGreaterThan(Sse2.And(lo, hi), zero)) != 0)
+                if (Vector128.BitwiseAnd(lo, hi).AsUInt64().ToScalar() != 0)
                     break;
 
-                Vector128<sbyte> eq2F = Sse2.CompareEqual(str, mask2F);
-                Vector128<sbyte> shift = Ssse3.Shuffle(lutShift, Sse2.Add(eq2F, hiNibbles));
+                Vector128<byte> eq2F = Vector128.Equals(str, mask2F);
+                Vector128<byte> shift = SimdShuffle(lutShift.AsByte(), Vector128.Add(eq2F, hiNibbles), f);
 
                 // Now simply add the delta values to the input:
-                str = Sse2.Add(str, shift);
+                str = Vector128.Add(str, shift);
 
                 // in, bits, upper case are most significant bits, lower case are least significant bits
                 // 00llllll 00kkkkLL 00jjKKKK 00JJJJJJ
                 // 00iiiiii 00hhhhII 00ggHHHH 00GGGGGG
                 // 00ffffff 00eeeeFF 00ddEEEE 00DDDDDD
                 // 00cccccc 00bbbbCC 00aaBBBB 00AAAAAA
 
-                Vector128<short> merge_ab_and_bc = Ssse3.MultiplyAddAdjacent(str.AsByte(), mergeConstant0);
+                Vector128<short> merge_ab_and_bc;
+                if (Ssse3.IsSupported)
+                {
+                    merge_ab_and_bc = Ssse3.MultiplyAddAdjacent(str.AsByte(), mergeConstant0.AsSByte());
+                }
+                else
+                {
+                    Vector128<ushort> evens = AdvSimd.ShiftLeftLogicalWideningLower(AdvSimd.Arm64.UnzipEven(str, one).GetLower(), 6);
+                    Vector128<ushort> odds = AdvSimd.Arm64.TransposeOdd(str, Vector128<byte>.Zero).AsUInt16();
+                    merge_ab_and_bc = Vector128.Add(evens, odds).AsInt16();
```

Comment:
Why this rather than the following?

```csharp
Vector128<ushort> tmp1 = AdvSimd.MultiplyWideningLower(str.AsByte(), mergeConstant0.AsByte());
Vector128<ushort> tmp2 = AdvSimd.MultiplyWideningUpper(str.AsByte(), mergeConstant0.AsByte());
merge_ab_and_bc = AdvSimd.AddPairwise(tmp1, tmp2).AsInt16();
```

This should be fine given `MultiplyAddAdjacent` is doing:
```csharp
short tmp1 = str.AsByte()[0], mergeConstant0.AsSByte()[0];
short tmp2 = str.AsByte()[1], mergeConstant0.AsSByte()[1];
return tmp1 + tmp2;
```

That is `(short)byte * (short)sbyte`. Since `mergeConstant0` is `0x40, 0x01` repeating and therefore is unsigned, it's fine to treat as `(ushort)byte * (ushort)byte` and then reinterpret to the signed result after the addition.

Reply:
I started with using multiplies, but then replaced them as they are fairly slow. Given it is only using constants, this seemed a better way of doing it. 
@TamarChristinaArm can probably give a more cycle accurate answer to this than me.

Reply:
It's likely worth measuring and seeing if the numbers show anything here

Reply:
> The latter could theoretically pipeline, but its not clear how probable that is and likely depends on alignment of the instructions relative to the loop (16 and 32-byte instruction windows are common on x86/x64 at least).

The OoO Window on big cores are extremely huge. The alignment of the loop doesn't matter as much here as you're not at the start of the loop.   Also the performance of `umull` fluctuates based on the core while permutes are always the same.  But yes, the goal here was to improve pipelining. The `umull` version will never pipeline on older cores where the permute version will likely do so.

> Zero is likely emitting a movi reg, #0 since there isn't a dedicated zero SIMD register and that may cost 2-cycles if there isn't a register renamer specially handling it.

On modern big cores, i.e. Cortex-A76. Neoverse-N1 etc it does get eliminated early, so it's 0 cost.

Reply:
>It's likely worth measuring and seeing if the numbers show anything here

On an Altra, with the code in this patch, Base64Decode is running at 0.59x of HEAD.

If I switch it to Multiply, Multiply, AddPairwise then we drop performance to 0.65x of HEAD.

So, definitely worth sticking with this patch.

I wonder if eventually this should be dragged into Vector128 as MultiplyAddAdjacent() ?